core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -157,6 +163,115 @@ public void segmentsLost(IntSet lostSegments) {
+            CompletableFuture<R> future = stage.toCompletableFuture();
+            // This will always be true if there isn't a store, however in most cases with a store this will
+            // be false as we have to wait until the store can publish all the entries (which is done asynchronously)
+            if (future.isDone() && !future.isCompletedExceptionally()) {
+               // If we complete the iteration try to remove the segment - so it can't be suspected
+               concurrentSegments.remove(nextSegment);
+               R notifiedValue;
+               // This segment was lost before we could complete our iteration - so we have to discard the result
+               if (listener.segmentsLost.contains(nextSegment)) {
+                  notifiedValue = null;
+               } else {
+                  notifiedValue = future.join();
+
+               }
+               completeTask(pendingCompletions, notifiedValue, processor);
+            } else {
+               // If we have a stage that isn't complete we have to convert to a serialized processor as multiple
+               // responses could come back at the same time
+               if (!serializedProcessor) {
+                  serializedProcessor = true;
+                  processor = processor.toSerialized();
+               }
+               final FlowableProcessor<R> processorToUse = processor;
+               stage.whenComplete((value, t) -> {
+                  if (t != null) {
+                     processorToUse.onError(t);
+                  } else {
+                     // If we complete the iteration try to remove the segment - so it can't be suspected
+                     concurrentSegments.remove(nextSegment);
+                     R notifiedValue;
+                     // This segment was lost before we could complete our iteration - so we have to discard the result
+                     if (listener.segmentsLost.contains(nextSegment)) {
+                        notifiedValue = null;
+                     } else {
+                        notifiedValue = value;
+                     }
+                     completeTask(pendingCompletions, notifiedValue, processorToUse);
+                  }
+               });
+            }
+         }
+         // Submitter has to complete a null value - signals that all segments have been submitted allowing the
+         // CompletionStage to actually complete
+         completeTask(pendingCompletions, null, processor);
+      } catch (Throwable t) {
+         processor.onError(t);
+      }
+   }
+
+   private static <V> void completeTask(AtomicInteger count, V value, FlowableProcessor<V> processor) {
+      if (value != null) {
+         processor.onNext(value);
+      }
+      if (count.decrementAndGet() == 0) {
+         processor.onComplete();
+      }
+   }
+
+   /**
+    * Retrieves the next int from the iterator which is thread safe from multiple invoking threads. This method
```

Comment:
I can reword the comment.  I am personally not keen on allocating an additional wrapper object, when the usage and documentation should be sufficient.

What do you mean about `concurrentSegments` ? That is already handled due to the fact that it is a `ConcurrentIntSet`.

Reply:
Like the IntelliJ warning says, "It is very difficult to guarantee correctness when such synchronization is used", because the javadoc is only on `getNextSegment`, and people who will modify the code later to access the iterator directly won't see it.

And yes, `concurrentSegments` is handled, but it's handled completely separate from the iteration, and I feel like they belong together. Although reading again I see `concurrentSegments` is already passed as a field in `SegmentListener`, with the comment "This variable should never be modified"

Reply:
> Like the IntelliJ warning says, "It is very difficult to guarantee correctness when such synchronization is used", because the javadoc is only on `getNextSegment`, and people who will modify the code later to access the iterator directly won't see it.

I can understand that more if it was a public method, but this is a private method only invoked by this class and does a very very simple job. It also has documentation stating its requirement regarding synchronization.

> And yes, `concurrentSegments` is handled, but it's handled completely separate from the iteration, and I feel like they belong together. Although reading again I see `concurrentSegments` is already passed as a field in `SegmentListener`, with the comment "This variable should never be modified"

That appears to be an outdated comment from probably 6 months ago when I original wrote the Publisher code that was then left in wait for 3 months until I got time to finish it. We modify the variable for rehash awareness and is an absolute requirement.